### PR TITLE
feat: Update support for segment analytics

### PR DESCRIPTION
**Pull request by benjackwhite**
## Changes

Fixes https://github.com/PostHog/posthog-js/issues/1106

Annoyingly Segment's documentation does not match reality. They claim that the Analytics JS will always have non-promise responses once it is ready but this isn't the case.

* Changes the Segment setup to all happen in one function
* Accounts for the User object being a promise or not
* Handles both distinct_id and device_id

## Checklist
- [x] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [ ] Accounted for the impact of any changes across different browsers
- [ ] Accounted for backwards compatibility of any changes (no breaking changes in posthog-js!)

### CODE REVIEW

File: src/extensions/segment-integration.ts [hunk 1]
```diff
@@ -34,23 +47,26 @@ interface SegmentPlugin {
...
+    load: (ctx: SegmentContext, instance: any, config?: any) => Promise<unknown>
+    unload?: (ctx: SegmentContext, instance: any) => Promise<unknown> | unknown
     ready?: () => Promise<unknown>
-    track?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
-    identify?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
-    page?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
-    group?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
-    alias?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
-    screen?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
+    track?: (ctx: SegmentContext) => Promise<SegmentContext> | SegmentContext
```
Comment (pauldambra) [inline src/extensions/segment-integration.ts:53]: nit that w could extract an interface here and this block of repeating lines would be easier on the eyes

**Review by pauldambra — APPROVED** [review]
🚢

**Review by robbie-c — APPROVED** [review]
LGTM

edit: lol too slow

**Comment by kamui** [comment]
Thank you!

**Pull request merged**